### PR TITLE
Add support of `dpnp.ndindex` class

### PR DESCRIPTION
**Pull request by antonwolfy**
The PR proposes to add `dpnp.ndindex` class. The implementation is fully leveraged on appropriate class in NumPy.
The tests are added to cover the new functionality.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: dpnp/dpnp_iface_indexing.py [hunk 1]
```diff
@@ -1059,6 +1060,57 @@ def mask_indices(
     return nonzero(a != 0)
 
 
+# pylint: disable=invalid-name
+ndindex = numpy.ndindex
+ndindex.__doc__ = """
+An N-dimensional iterator object to index arrays.
+
+Given the shape of an array, an `ndindex` instance iterates over the
```
Comment (vtavana) [inline dpnp/dpnp_iface_indexing.py:1068]:
```suggestion
Given the shape of an array, an :obj:`dpnp.ndindex` instance iterates over the
```

**Review by vtavana — COMMENTED** [review]
The documentation shows a few additional methods `__eq__`, `__ne__`, etc. Are they relevant? Is it possible to remove them?
https://intelpython.github.io/dpnp/pull/2157/reference/generated/dpnp.ndindex.html#dpnp.ndindex.__eq__

File: dpnp/dpnp_utils/dpnp_utils_pad.py [hunk 2]
```diff
@@ -642,8 +642,7 @@ def dpnp_pad(array, pad_width, mode="constant", **kwargs):
 
             # compute indices for the iteration axes, and append a trailing
             # ellipsis to prevent 0d arrays decaying to scalars
-            # TODO: replace with dpnp.ndindex when implemented
```
Comment (vtavana) [inline dpnp/dpnp_utils/dpnp_utils_pad.py:645]: There is another TODO for replacing `numpy.ndindex` with `dpnp.ndindex` in 
`dpnp_iface_functional.py`.
https://github.com/IntelPython/dpnp/blob/master/dpnp/dpnp_iface_functional.py#L154

**Comment by antonwolfy** [comment]
> The documentation shows a few additional methods `__eq__`, `__ne__`, etc. Are they relevant? Is it possible to remove them? https://intelpython.github.io/dpnp/pull/2157/reference/generated/dpnp.ndindex.html#dpnp.ndindex.__eq__

The issue will be resolved by separate PR: #2187

**Review by vtavana — APPROVED** [review]
Thank you, @antonwolfy! Documentation looks fine now.

**Pull request merged**